mlir/lib/Dialect/DLTI/DLTI.cpp

```diff
@@ -424,8 +424,8 @@ getClosestQueryable(Operation *op) {
   return std::pair(queryable, op);
 }
 
-FailureOr<Attribute> dlti::query(Operation *op, ArrayRef<StringAttr> keys,
-                                 bool emitError) {
+FailureOr<Attribute>
+dlti::query(Operation *op, ArrayRef<DataLayoutEntryKey> keys, bool emitError) {
```

Comment:
Unrelated to this PR so feel free to ignore for now but just came to my mind.

What's the desired behavior when `keys` is empty? Right now it works but it's not covered by any test.

Reply:
At the moment an empty `keys` would return the DLTI attribute that is closest to `op`, if such an attribute exists. This behaviour might be useful/desirable - you could then work directly with that attribute, for example. 

A reason this behaviour might be less desirable is that we would like to implement shadowing, i.e. `dlti::query` would in try all DLTI attributes that are attached to ancestors of `op` in turn, until one can provide a value for `keys`. In that case an empty `keys` would more naturally correspond to a queryable object which would provide the same lookup capability, i.e. would delegate key look ups to the same underlying attributes.

In the absence of a strong argument for what should be returned, I am inclined to just disallow empty `keys` for now. I am fine with doing it in this PR, or else I would address it when I do the shadowing work.

Let me know what you feel would be best.

Reply:
Shadowing use case is interesting but let's leave that for other time.

Intuitively and based on the API docs, I would expect it to fail on no keys i.e., nothing to match equals no match.
But a special case where empty key means returning DLTI root (or closest or w/e) is IMO also fine as long as it is documented.

Either way, it'd be great if you could add one test case for `transform.dlti.query []` to ensure this is handled at all.

Reply:
> At the moment an empty `keys` would return the DLTI attribute that is closest to `op`, if such an attribute exists. This behaviour might be useful/desirable - you could then work directly with that attribute, for example.

I would not go there. Dictionaries do not guarantee order and you could have inconsistent behaviour across runs. I would right now restrict what we can put in keys and have tests for the cases where we care about and negative tests for the ones we know could work but we restrict to avoid complexity.